src/main/java/duke/parser/Parser.java

```diff
@@ -0,0 +1,155 @@
+package duke.parser;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import duke.command.AddCommand;
+import duke.command.Command;
+import duke.command.DeleteCommand;
+import duke.command.ExitCommand;
+import duke.command.FindCommand;
+import duke.command.ListCommand;
+import duke.command.MarkCommand;
+import duke.command.UnmarkCommand;
+import duke.exceptions.DukeException;
+import duke.tasklist.TaskList;
+import duke.tasktypes.Deadline;
+import duke.tasktypes.Event;
+import duke.tasktypes.Task;
+import duke.tasktypes.ToDo;
+
+
+/**
+ * Represents an input Parser.
+ * A Parser makes sense of User input and decides Duke's follow-up action.
+ */
+public class Parser {
+
+    private static final String[] VALUE_COMMANDS = {"unmark ", "mark ", "delete ", "find "};
+    private static final String[] TASK_COMMANDS = {"deadline ", "todo ", "event "};
+
+    public Parser() {
+    }
+
+    /**
+     * Returns a Command object after parsing User input.
+     *
+     * @param fullCommand String representation of User input.
+     * @param tasks TaskList instance containing existing tasks.
+     * @return Command object that determines follow-up action.
+     * @throws DukeException If there are discrepancies in User input.
+     */
+    public static Command parse(String fullCommand, TaskList tasks) throws DukeException {
+        if (fullCommand.equals("bye")) {
+            return new ExitCommand();
+        } else if (fullCommand.equals("list")) {
+            return new ListCommand();
+        }
+
+        String command = Parser.returnCommand(fullCommand, VALUE_COMMANDS);
+        String[] inputArr;
+        switch (command) {
+        case "unmark ":
+            inputArr = fullCommand.split(" ");
+            int toUnmark = Integer.parseInt(inputArr[1]);
+            if (toUnmark > tasks.getNumTasks()) {
+                throw new DukeException("Task does not exist! Please enter valid input!");
```

Comment:
Perhaps you can extract the different `Exceptions` into different classes? 

Suggestion would be:
```
Exception (Base Class)
|_ InvalidArgumentException
|_ MissingArgumentException
|_ InvalidDateException
|_ (etc)
```

Reply:
Good suggestion. I will consider doing that!